### PR TITLE
fix(bin/client): don't close closing connection

### PR DESCRIPTION
**Pull request by mxinden**
The `bin/src/client/mod.rs` `Runner::run` function continuously checks whether there is more work. In case there is none, it initiates closing of the connection (`self.client.close`) and then `continue`s to the top of the loop in order to send out a closing frame.

https://github.com/mozilla/neqo/blob/14cafbaa7fa88434def2c1d19e932c08e00173f8/neqo-bin/src/client/mod.rs#L376-L409

There is a potential busy loop when closing an already closing connection. `Runner::run` will call `self.client.close` and then continously `continue` to the top of the loop.

This commit differentiates a connection state in `NotClosing`, `Closing` and `Closed`. It only attempts to close a `NotClosing` connection and only then `continue`s to the top of the loop.

---

Fixes https://github.com/mozilla/neqo/issues/1864.

### CODE REVIEW

**Comment by larseggert** [comment]
Any chance we can refactor this to avoid the duplication between h09 and h3?

File: neqo-bin/src/client/http3.rs [hunk 1]
```diff
@@ -106,15 +106,22 @@ pub(crate) fn create_client(
 }
 
 impl super::Client for Http3Client {
-    fn is_closed(&self) -> Result<bool, ConnectionError> {
-        match self.state() {
-            Http3State::Closed(
-                ConnectionError::Transport(neqo_transport::Error::NoError)
+    fn is_closed(&self) -> Result<CloseState, ConnectionError> {
```
Comment (martinthomson) [inline neqo-bin/src/client/http3.rs:109]: It's a shame that this is almost completely duplicated code.

**Review by martinthomson — APPROVED** [review]

**Comment by mxinden** [comment]
@larseggert @martinthomson the latest commit does the following:

- Introduce `ConnectionError::is_error`. (Long term I think it would be great for `ConnectionError` not to contain `ApplicationError(0)` and `TransportError::NoError`, but I think that is a matter for another pull request.)
- Leverage `TryFrom`, making the conversion a bit more idiomatic.

I think it is an improvement. I don't think it is great.

Let me know what you think.

**Comment by martinthomson** [comment]
@mxinden 

> Long term I think it would be great for ConnectionError not to contain ApplicationError(0) and TransportError::NoError, but I think that is a matter for another pull request.

That might be because `ConnectionError` is not a good name for the type.  A better name might be `CloseReason`.

**Pull request merged**